crates/turborepo-filewatch/src/package_watcher.rs

```diff
@@ -186,32 +244,56 @@ impl Subscriber {
         };
 
-        // state represents our current understanding of the underlying filesystem, and
-        // is expected to be mutated in place by handle_file_event. Both
-        // rediscover_everything and handle_file_event are responsible for
-        // broadcasting updates to state.
-        let mut state = self.rediscover_and_write_state().await;
+        let (package_state_tx, mut package_state_rx) = mpsc::channel::<DiscoveryResult>(256);
+
+        let (version, debouncer) = self.queue_rediscovery(true, package_state_tx.clone());
+
+        // state represents the current state of this process, and is expected to be
+        // updated in place by the various handler functions.
+        let mut state = State::Pending { debouncer, version };
 
         tracing::debug!("package watcher ready {:?}", state);
         loop {
-            let file_event = recv.recv().await;
-            match file_event {
-                Ok(Ok(event)) => self.handle_file_event(&mut state, &event).await,
-                // if we get an error, we need to re-discover the packages
-                Ok(Err(_)) => state = self.rediscover_and_write_state().await,
-                Err(e @ RecvError::Closed) => {
-                    return PackageWatcherProcessError::FilewatchingClosed(e)
-                }
-                // if we end up lagging, warn and rediscover packages
-                Err(RecvError::Lagged(count)) => {
-                    tracing::warn!("lagged behind {count} processing file watching events");
-                    state = self.rediscover_and_write_state().await;
+            select! {
+                Some(discovery_result) = package_state_rx.recv() => {
+                    self.handle_discovery_result(discovery_result, &mut state);
+                },
+                file_event = recv.recv() => {
+                    match file_event {
+                        Ok(Ok(event)) => self.handle_file_event(&mut state, &event, &package_state_tx).await,
+                        // if we get an error, we need to re-discover the packages
+                        Ok(Err(_)) => self.bump_or_queue_rediscovery(&mut state, &package_state_tx),
+                        Err(e @ RecvError::Closed) => {
+                            return PackageWatcherProcessError::FilewatchingClosed(e)
+                        }
+                        // if we end up lagging, warn and rediscover packages
+                        Err(RecvError::Lagged(count)) => {
+                            tracing::warn!("lagged behind {count} processing file watching events");
+                            self.bump_or_queue_rediscovery(&mut state, &package_state_tx);
+                        }
+                    }
                 }
             }
             tracing::trace!("package watcher state: {:?}", state);
         }
     }
 
+    fn handle_discovery_result(&self, package_result: DiscoveryResult, state: &mut State) {
```

Comment:
What should the behaviour be if we get a `DiscoveryResult` with valid `State`? Should our state perhaps also store a version and be updated?